raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -470,15 +470,53 @@ public void transitionToUnattachedVotedState(
      */
     public void transitionToFollower(int epoch, int leaderId, Endpoints endpoints) {
         int currentEpoch = state.epoch();
-        if (localId.isPresent() && leaderId == localId.getAsInt()) {
-            throw new IllegalStateException("Cannot transition to Follower with leader " + leaderId +
-                " and epoch " + epoch + " since it matches the local broker.id " + localId);
+        if (endpoints.isEmpty()) {
```

Comment:
is it possible this might actually be empty? (leader on version prior to https://github.com/apache/kafka/pull/16235/files#)

Reply:
Yes. It can be empty if the leader is on the old software version and sends version 0 of the BEGIN_QUORUM_EPOCH request.